packages/react-shared/src/components/link.tsx

```diff
@@ -0,0 +1,101 @@
+import * as _ from 'lodash';
+import * as React from 'react';
+import { Link as ReactRounterLink } from 'react-router-dom';
+import { Link as ScrollLink } from 'react-scroll';
+
+import { LinkType } from '../types';
+import { constants } from '../utils/constants';
+
+export interface LinkProps {
+    to: string;
+    type?: LinkType;
+    shouldOpenInNewTab?: boolean;
+    style?: React.CSSProperties;
+    className?: string;
+    onMouseOver?: (event: React.MouseEvent<HTMLElement>) => void;
+    onMouseLeave?: (event: React.MouseEvent<HTMLElement>) => void;
+    onMouseEnter?: (event: React.MouseEvent<HTMLElement>) => void;
+    containerId?: string;
+}
```

Comment:
nit: Some props like `className` or `containerId` are ignored if a certain `type` prop is used. 

You could do something like (I think? This works in Flow):
```
interface CommonProps {}
interface ScrollLinkProps {
  type: LinkType.ReactRoute;
  containerId?: string
}
interface ExternalLinkProps {
  type: LinkType.External;
  className?: string;
}

export type LinkProps = CommonProps & (ExternalLinkProps | ScrollLinkProps);
```
Then when you set the type prop to something the type system can infer what other props are usable.

Also is the `style` prop necessary?

Reply:
Hm, tried your approach but it was giving me issues with the `defaultProps` instance object. I ended up figuring out a way to get all props to apply to all link types, **except** for `shouldOpenInNewTab` & `linkType.ReactScroll`, which should never be passed in together. For that one, I added an assert that throws. This type of mistake will now be caught during development.

Reply:
I went ahead and removed the `style` prop.